nfsv4/client.c

```diff
@@ -110,32 +114,39 @@ int nfs4_append(nfs4_file f, void *source, bytes length)
     p.mask = NFS4_PROP_SIZE;
     p.size = f->expected_size;
     // get the offset in case we fail
-    push_op(r, OP_GETATTR, set_expected_size, &f);    
+    push_op(r, OP_GETATTR, set_expected_size, &f);
     push_fattr_mask(r, NFS4_PROP_SIZE);
     push_op(r, OP_VERIFY, 0, 0);
     push_fattr(r, &p);
     push_op(r, OP_SETATTR, parse_attrmask, 0);
-    push_stateid(r, &f->open_sid);    
+    push_stateid(r, &f->open_sid);
     push_fattr(r, &p);    
     push_lock(r, &f->open_sid, WRITE_LT, f->expected_size, f->expected_size + length, &f->latest_sid);
     buffer b = alloca_wrap_buffer(source, length);
     u64 offset = f->expected_size;
-        
+ 
+    transact(r);
+
     while (buffer_length(b)) {
-        if (!r) r = file_rpc(f);
+        r = file_rpc(f);
         // join
         offset += push_write(r, offset, b, &f->latest_sid);
         rpc_send(r);
-        r = 0;
     }
     // drain
     
     r = file_rpc(f);
     push_unlock(r, &f->latest_sid, WRITE_LT, f->expected_size, f->expected_size + length);
-    rpc_send(r);    
     // fix direct transmission of nfs4 error codes
     status s = transact(r);
-    return api_check(f->c, s);
+    if (nfs4_is_error(s)) {
```

Comment:
I think the `if` part of this clause is implemented by `api_check`,
storing into error_string and returning -1,
though it looks like nfs_error_num is not updated

Reply:
as a detail for another PR, perhaps `api_check` should store into `nfs_error_num` as well. I think `nfs_error_num` was a field I added in a couple months ago

Reply:
cool I will modify the api_check, rewrite this part